Detectors/TPC/simulation/include/TPCSimulation/ClusterContainer.h

```diff
@@ -44,18 +38,21 @@ namespace o2 {
       /// @param padsigma Sigma of cluster in pad direction
       /// @param timemean Mean position of cluster in time direction
       /// @param timesigma Sigma of cluster in time direction
-      Cluster* AddCluster(Int_t cru, Int_t row, Float_t qTot, Float_t qMax,
-			  Float_t pad, Float_t time, Float_t sigmapad,
-			  Float_t sigmatime);
-
-      // Copy container info into the output container
-      void FillOutputContainer(TClonesArray *outputcont);
-
-      Int_t GetEntries() { return mClusterArray->GetEntries(); };
-
-    private:
-      Int_t         mNclusters;        // number of clusters
-      TClonesArray* mClusterArray;      // array for clusters
+      template<typename ClusterType>
+      static ClusterType* AddCluster(std::vector<ClusterType> *output,
+		             Int_t cru, Int_t row, Float_t qTot, Float_t qMax,
+			     Float_t meanpad, Float_t meantime, Float_t sigmapad,
+			     Float_t sigmatime) {
+        assert(output);
+        output->emplace_back(); // emplace_back a defaut constructed cluster of type ClusterType
+        auto& cluster = output->back();
+        // set its concrete parameters:
+        // ATTENTION: the order of parameters in setParameters is different than in AddCluster!
+        cluster.setParameters(cru, row, qTot, qMax,
+                              meanpad, sigmapad,
+                              meantime, sigmatime);
+        return &cluster;
+      }
```

Comment:
In this class it seem there are still many function name to following the naming conventions (starting with a capital letter). Should this be cleaned up? I could do it, or we wait till after merging this.

Reply:
After this PR please! In principle we should wait until the corresponding code checker is in place.

Reply:
Fine with me

Reply:
just a question: the first argument here (in AddCluster) is an input/output type where you expect the object to be valid - would a ref type argument not communicate this in a clearer way than a pointer? Additionally, the assert is optional (you can compile with assertions disabled), so it might be a problem down the road...

Reply:
@mkrzewic : This is of course generally preferred. An excuse is that this is in principle an orthogonal refactoring which has nothing to do with the features of this PR. I'll open a ticket ...-:)